auth.go

```diff
@@ -45,14 +48,52 @@ func validateSession(auth authentication) error {
 	if err := validateAuth(Salesforce{auth: &auth}); err != nil {
 		return err
 	}
-	_, err := doRequest(http.MethodGet, "/limits", jsonType, auth, "")
+	_, err := doRequest(&auth, requestPayload{
+		method:  http.MethodGet,
+		uri:     "/limits",
+		content: jsonType,
+	})
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
 
+func refreshSession(auth *authentication) error {
+	var refreshedAuth *authentication
+	var err error
+
+	switch grantType := auth.grantType; grantType {
+	case grantTypeClientCredentials:
+		refreshedAuth, err = clientCredentialsFlow(
+			auth.creds.Domain,
+			auth.creds.ConsumerKey,
+			auth.creds.ConsumerSecret,
+		)
```

Comment:
I would have returned the error when present in both branches I think.

Even if your code is OK, checking the err closer would be more "classic"